rsyncwrap/main.py

```diff
@@ -7,27 +7,56 @@
 from dataclasses import dataclass, field
 from decimal import Decimal
 from pathlib import Path
-from typing import Dict, Iterator, List, Optional, Sequence, Union
+from typing import Dict, Iterator, List, Tuple, Optional, Sequence, Union
 
-from rsyncwrap.helpers import parts_to_path
+from rsyncwrap.helpers import is_remote, parse_location, parts_to_path
 
 try:
     from functools import cached_property
 except ImportError:
     from rsyncwrap.helpers import cached_property
 
 
-def _rsync(source_paths: Sequence[Path], dest_path: Path) -> Iterator[Union[str, int]]:
+def _rsync(
+    source_locations: Sequence[str], dest_location: str, ssh_config: Dict, dry_run: bool
+) -> Iterator[Union[str, int]]:
     """
     Runs rsync and yields lines of output.
     """
-    cmd = [
-        "rsync",
-        "-a",
-        "--progress",
-        *[str(p) for p in source_paths],
-        str(dest_path),
-    ]
+    # If destination is local, make sure dir exists
+    dest_user, dest_remote, dest_path = parse_location(dest_location)
+    if not is_remote(dest_location):
+        assert dest_path.is_dir(), f"{dest_path} is not a directory."
+
+    # Make sure local sources, if any, must exist
+    for l in source_locations:
+        if not is_remote(l):
+            path = parse_location(l)[-1]
+            assert path.exists(), f"Local source '{path}' does not exist."
+
+    # Build list of SSH config options
+    ssh = [f"-o {k}={v}" for k, v in ssh_config.items()]
+
+    # Build command line. Start with basic rsync
+    cmd = ["rsync", "--archive", "--progress"]
+    # Configure dry run
+    if dry_run:
+        cmd.append("--dry-run")
+    # Add SSH specific config
+    if ssh:
+        ssh.insert(0, "-e ssh")
+        cmd.append(" ".join(ssh))
+    # Add sources
+    cmd.append(*[str(p) for p in source_locations])
+    # Add destination
+    cmd.append(dest_location)
+
+    # Char ";" is invalid for security. Is it handled elsewhere by Python magic?
+    # Note: := is new in Python 3.8
+    if ";" in (l := " ".join(cmd)):
+        raise ValueError(f"Command line contains ';': '{l}'")
+    #print(cmd)
```

Comment:
I suppose this and the next line can be taken out.

Reply:
Right, with subprocess run, there seems to not be a way to use a ";" to execute more than one command. Do you agree?